### PR TITLE
DAOS-623 cq: speedup get_release_branch

### PR DESCRIPTION
**Pull request by daltonbohning**
Speedup get_release_branch by filtering release/* branches to only 2.4+

Skip-test: true
Skip-tests: true
Skip-unit-tests: true
Skip-fault-injection-test: true

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by daltonbohning** [comment]
@brianjmurrell On my boro node, the current version of this takes ~0.8s
```
$ git branch -r | sed -ne "/^  $origin\\/release\\/[0-9]/s/^  $origin\\///p"
release/0.9
release/1.0
release/1.2
release/2.0
release/2.2
release/2.4

$ time /home/dbohning/daos/utils/rpms/packaging/get_release_branch
master

real    0m0.811s
```


Whereas this updated version takes ~0.1s
```
$ git branch -r | sed -ne "/\(^  $origin\\/release\\/2.[4-9]\|^  $origin\\/release\\/[3-9]\)/s/^  $origin\\///p"
release/2.4

$ time /home/dbohning/daos/utils/rpms/packaging/get_release_branch
master

real    0m0.114s
```
And since the githooks run this three times: for commit-msg, pre-commit, prepare-commit-msg,
This (on my boro node) reduces the time to detect the target branch from `0.8*3=2.4s -> 0.1*3=0.3s`

And timing a simple commit
```
$ time git commit -sa --amend --no-edit
...
real    0m2.520s
```
Which means previously detecting the target branch dominated the githook time

**Comment by brianjmurrell** [comment]
> @brianjmurrell On my boro node, the current version of this takes ~0.8s

But never, ever needs to be updated for newer daos releases.

> Whereas this updated version takes ~0.1s
> 
> ```
> $ git branch -r | sed -ne "/\(^  $origin\\/release\\/2.[4-9]\|^  $origin\\/release\\/[3-9]\)/s/^  $origin\\///p"
> release/2.4
> 
> $ time /home/dbohning/daos/utils/rpms/packaging/get_release_branch
> master
> 
> real    0m0.114s
> ```

But only works until we hit 3.x releases and then needs updating (and gets progressively slower again as we proceed through more 2.x releases without updating the lower bound of the `[4-9]` constraint).  Who is going [to remember to] do that?

Not arguing with the potential speedup, just saying that it comes at a cost of future maintenance, potentially (likely even) only after somebody wastes^Wspends a bunch of time figuring out that it didn't get bumped when the 3.x release stream starts like it needed to.

**Comment by daltonbohning** [comment]
> But never, ever needs to be updated for newer daos releases.

> But only works until we hit 3.x releases and then needs updating (and gets progressively slower again as we proceed through more 2.x releases without updating the lower bound of the `[4-9]` constraint). Who is going [to remember to] do that?
> 
> Not arguing with the potential speedup, just saying that it comes at a cost of future maintenance, potentially (likely even) only after somebody wastes^Wspends a bunch of time figuring out that it didn't get bumped when the 3.x release stream starts like it needed to.

My change will match release/2.4+ and release/3+, including release/4*, release/5*, etc.

**Comment by daltonbohning** [comment]
> gets progressively slower again as we proceed through more 2.x releases without updating the lower bound of the [4-9] constraint

True, but this is just an optimization. Worst case, we forget about this forever and it just gets slow. Best case, we remember and can optimize it again

**Comment by daltonbohning** [comment]
I definitely agree we don't want to be forced to update this or risk silent breakage. So e.g. it does match future releases
```
$ origin=origin
$ echo "
  origin/release/2.0
  origin/release/2.2
  origin/release/2.4
  origin/release/2.6
  origin/release/2.8
  origin/release/3.0
  origin/release/3.2
  origin/release/4.0
" | sed -ne "/\(^  $origin\\/release\\/2.[4-9]\|^  $origin\\/release\\/[3-9]\)/s/^  $origin\\///p"

release/2.4
release/2.6
release/2.8
release/3.0
release/3.2
release/4.0
```

**Comment by brianjmurrell** [comment]
> I definitely agree we don't want to be forced to update this or risk silent breakage. So e.g. it does match future releases
> 
> ```
> $ origin=origin
> $ echo "
>   origin/release/2.0
>   origin/release/2.2
>   origin/release/2.4
>   origin/release/2.6
>   origin/release/2.8
>   origin/release/3.0
>   origin/release/3.2
>   origin/release/4.0
> " | sed -ne "/\(^  $origin\\/release\\/2.[4-9]\|^  $origin\\/release\\/[3-9]\)/s/^  $origin\\///p"
> 
> release/2.4
> release/2.6
> release/2.8
> release/3.0
> release/3.2
> release/4.0
> ```

Oh.  I missed what the rest of the RE was doing.

File: utils/rpms/packaging/get_release_branch [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+# base branches can be master, release/2.4+, release/3+
+# or optionally branches passed into $1
 
 set -eu -o pipefail
 IFS=' ' read -r -a add_bases <<< "${1:-}"
 origin=origin
 mapfile -t all_bases < <(echo "master"
-                         git branch -r | sed -ne "/^  $origin\\/release\\/[0-9]/s/^  $origin\\///p")
+                         git branch -r | sed -ne "/\(^  $origin\\/release\\/2.[4-9]\|\
+                                                     ^  $origin\\/release\\/[3-9]\)/s/^  $origin\\///p")
```
Comment (brianjmurrell) [inline utils/rpms/packaging/get_release_branch:12]: The RE can be reduced down to a more clear and concise:

```suggestion
                         git branch -r | sed -ne "/^  $origin\\/release\\/(2.[4-9]\|[3-9]/s/^  $origin\\///p")
```

**Review by brianjmurrell — APPROVED** [review]

**Review by phender — APPROVED** [review]

**Pull request merged**